### PR TITLE
Sih tailsitter

### PR DESCRIPTION
**Pull request by romain-chiap**
Tailsitter model now supported in SIH!

I'm still working on a nice rending for jMAVSim...

Model based on my Master thesis
"Chiappinelli, Romain. Modeling and control of a flying wing tailsitter unmanned aerial vehicle. McGill University (Canada), Masters Thesis, 2018."

### CODE REVIEW

**Comment by xdwgood** [comment]
nice！！！

**Comment by xdwgood** [comment]
how to run it? maybe: `make px4_sitl_default jamvsim_tailsitter`?

**Comment by xdwgood** [comment]
how to run it?
I want to have a try.
Could you provide specific steps? I tried the following steps but failed. 
1:updated firmware (branch:sih-ts)
2:open qgc and switch `sih plane aert`
3:`make px4_sitl_default`
4:`./Tools/jmavsim_run.sh -q -ts -d /dev/ttyACM0 -b 2000000 -o`

File: ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# - without safety switch
+param set-default COM_PREARM_MODE 0
+param set-default CBRK_IO_SAFETY 22027
+
+param set-default BAT_N_CELLS 3
+
+param set SIH_T_MAX 2.0
+param set SIH_Q_MAX 0.0165
+param set SIH_MASS 0.2
+param set SIH_IXX 0.00354 	# IXX and IZZ are inverted from the thesis as the body frame is pitched by 90 deg
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil:51]: Can you put comments on a new line so they're stripped from the binary?
```suggestion
# IXX and IZZ are inverted from the thesis as the body frame is pitched by 90 deg
param set SIH_IXX 0.00354
```

File: ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+param set SIH_IYY 0.000625
+param set SIH_IZZ 0.00300
+param set SIH_IXZ 0.0
+param set SIH_KDV 0.2
+param set SIH_L_ROLL 0.145
+
+param set SIH_VEHICLE_TYPE 2 	# sih as tailsitter
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil:58]: Can  you update the parameter metadata as well?

File: src/modules/sih/aero.hpp [hunk 3]
```diff
@@ -262,6 +271,12 @@ class AeroSeg
 	// must be called after update_aero()
 	matrix::Vector3f get_Ma() const { return _Ma; }
 
+	// return the velocity in segment frame
+	matrix::Vector3f get_vS() const { return _v_S; }
+
+	// copy assignment operator
+	AeroSeg& operator=(const AeroSeg&){return *this;}
```
Comment (bkueng) [inline src/modules/sih/aero.hpp:278]: This does not actually assign.

Comment (romain-chiap) [reply src/modules/sih/aero.hpp:278]: we can leave that, I'm finally not using it.

File: ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil (hunk 2 above)
Comment (romain-chiap) [reply ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil]: I ran `make parameters_metadata`, added, committed, and pushed. Is that enough?

I think they are generated inside ./build which is not committed.

Comment (bkueng) [reply ROMFS/px4fmu_common/init.d/airframes/1102_tailsitter_duo_sih.hil]: Sorry, what I meant was to add tailsitter here: https://github.com/PX4/PX4-Autopilot/blob/c60f0934e94a401ca1d13b181c76143114e5f510/src/modules/sih/sih_params.c#L443

**Comment by romain-chiap** [comment]
There was a roll-pitch-yaw coupling during MC altitude mode. I reduced it by saturating the mixer. It is still present during aggressive maneuvers, but I'm satisfied with it.

Can we merge?

**Review by bkueng — APPROVED** [review]
There's a conflict with the Tools/jMAVSim submodule. Can you resolve it?
Otherwise it's good to merge.

**Comment by romain-chiap** [comment]
I did a 'git rebase master' and now all the commits are part of this branch. I don't know if it's an issue in itself.
Also, the sih doesn't work now. **The sih doesn't work either on master**. 

The sih reads the topic 'actuator_outputs', which seem to not be modulated in master.
It almost behave like if the mixer did not start properly. (I know there were modifications of the mixers recently). @dagar any clue on that?

**Comment by bkueng** [comment]
Something is off now, can you check the commits? Can you do `git fetch origin && git rebase origin/master`?

> The sih reads the topic 'actuator_outputs', which seem to not be modulated in master.
It almost behave like if the mixer did not start properly. (I know there were modifications of the mixers recently). @dagar any clue on that?

That should still behave the same. How did you run it?

**Comment by NicolasM0** [comment]
> I did a 'git rebase master' and now all the commits are part of this branch. I don't know if it's an issue in itself. Also, the sih doesn't work now. **The sih doesn't work either on master**.
> 
> The sih reads the topic 'actuator_outputs', which seem to not be modulated in master. It almost behave like if the mixer did not start properly. (I know there were modifications of the mixers recently). @dagar any clue on that?

I opened an issue for that: https://github.com/PX4/PX4-Autopilot/issues/18668
Setting UAVCAN_ENABLE to 0 fix it

**Comment by romain-chiap** [comment]
Thanks Nicolas! 
@bkueng do you want me to undo the `rebase` ?

**Comment by romain-chiap** [comment]
can we merge this?
I believe we'll never us UAVCAN during sih? will we?

**Comment by bkueng** [comment]
I don't think it's mergable in the current state. The changes looked good, but please fix the commits. Alternatively you can create a new branch off PX4 master and cherry-pick the commits.

**Pull request closed without merging**